pkg/cmd/docgen/diagrams.go

```diff
@@ -1526,6 +1540,68 @@ var specs = []stmtSpec{
 	},
 }
 
+// getAllStmtSpecs returns a slice of stmtSpecs for all sql.y statements that
+// should have a diagram generated for.
+// getAllStmtSpecs appends to the "specs" slice any sql.y statements that do
+// not have an entry in specs but are not specified to be skipped.
+func getAllStmtSpecs(sqlGrammarFile string, bnfAPITimeout time.Duration) ([]stmtSpec, error) {
+	sqlStmts := make(map[string]struct{})
+	// Map all the sql stmts that are defined in specs.
+	for _, s := range specs {
+		sqlStmts[s.GetStatement()] = struct{}{}
+	}
+
+	file, err := os.Open(sqlGrammarFile)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	bnf, err := runBNF(sqlGrammarFile, bnfAPITimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	br := func() io.Reader {
+		return bytes.NewReader(bnf)
+	}
+
+	grammar, err := extract.ParseGrammar(br())
+	if err != nil {
+		return nil, err
+	}
+
+	scanner := bufio.NewScanner(file)
+	stmtRegex, err := regexp.Compile(`%type <tree.Statement>`)
+	if err != nil {
+		return nil, err
+	}
+	for scanner.Scan() {
+		text := scanner.Text()
+		if stmtRegex.MatchString(text) {
```

Comment:
i would do `var stmtRegex = regexp.MustCompile("%type\s*<tree.Statement>\s*(.*)$")` as a global, then instead of strings.Split use regexp groupings:

```
var stmtRegex = regexp.MustCompile("%type\s*<tree.Statement>\s*(.*)$")

...

for scanner.Scan() {
   text := scanner.Text()
   if matches := stmtRegex. FindAllStringSubmatch(text, -1); len(matches) > 0 {
      stmts := strings.Split(matches[1], " ")

      for _, stmt := range stmts {
            // etc
       }
  }
}
```